DUnitX.TestDataProvider.pas

```diff
@@ -0,0 +1,86 @@
+(*
+****************************************************************
+* Unit:		      DUnitX.TestDataProvider
+* Area:	      	TestCases
+* Author:	      Uwe Rupprecht
+* Purpose:      Base Class and Factory for TestDataProvider
+* Version:	    1.0.0.0
+* Last Change:	16.08.2018 15:25
+* History:      -
+****************************************************************
+* This Source Code Form is subject to the terms of the
+* Mozilla Public License, v. 2.0. If a copy of the MPL was not
+* distributed with this file, You can obtain one at
+* https://mozilla.org/MPL/2.0/.
+****************************************************************
+*)
+unit DUnitX.TestDataProvider;
+
+interface
+uses
+  System.Classes,
+  System.Generics.Collections,
+  DUnitX.Types,
+  DUnitX.InternalInterfaces;
+
+TYPE
+  TTestDataProviderBase = Class abstract (TInterfacedObject,ITestDataProvider)
+    private
+    protected
+    public
+      Constructor Create;virtual;Abstract;
+      function GetCaseAmount(Methodname:string):integer;Virtual;Abstract;
+      function GetCaseName(Methodname:string):String;Virtual;Abstract;
+      function GetCaseParams(Methodname:string;casenr:integer):TValuearray;Virtual;Abstract;
+  End;
+  TTestDataProviderBaseClass = class of TTestDataProviderbase;
+
+  TestDataProviderManager = Class
```

Comment:
If the attribute took a class parameter rather than a string, then this wouldn't be needed, as all this is doing is looking up a class type by a string.

Reply:
I'm still not sure why the TestDataProviderManager class is needed. By changing the attribute to take the class (I see you have added a second constructor), you have all the information you need to create the provider (the class type). If you remove the constructor that takes the string, the TestDataProviderManager is not needed at all as there is no lookup needed.

I had though perhaps the TestDataProviderManager would be caching the provider instances, since the providers take no parameters, if there is any expense in creating the provider (like parsing a large json file), then perhaps the TestDataProviderManager could stay and cache any providers it creates? At the monent the TestDataProviderManager is purely a dictionary to lookup a class by name, and if you use the attribute constructor with the class type parameter, that lookup is not needed.